dashboard/master.py

```diff
@@ -0,0 +1,153 @@
+            results.append(node_info)
+
+        return results
+
+    async def _update_nodes(self):
+        while True:
+            try:
+                nodes = await self._get_nodes()
+                node_ips = [node["nodeManagerAddress"] for node in nodes]
+
+                agents = {}
+                for node in nodes:
+                    node_ip = node["nodeManagerAddress"]
+                    if node_ip not in DataSource.agents:
+                        key = "{}{}".format(
+                            dashboard_consts.DASHBOARD_AGENT_PORT_PREFIX,
+                            node_ip)
+                        agent_port = await self.aioredis_client.get(key)
+                        if agent_port:
+                            agents[node_ip] = agent_port
+
+                DataSource.agents.reset(agents)
+                DataSource.nodes.reset(
+                    dict(
+                        zip(node_ips,
+                            [node["nodeManagerPort"] for node in nodes])))
+                DataSource.hostname_to_ip.reset(
+                    dict(
+                        zip([node["nodeManagerHostname"] for node in nodes],
+                            node_ips)))
+                DataSource.ip_to_hostname.reset(
+                    dict(
+                        zip(node_ips,
+                            [node["nodeManagerHostname"] for node in nodes])))
+            except Exception as ex:
+                logger.exception(ex)
+            finally:
+                await asyncio.sleep(
+                    dashboard_consts.UPDATE_NODES_INTERVAL_SECONDS)
+
+    def _load_modules(self):
+        """Load dashboard master modules."""
+        modules = []
+        for cls in self._master_cls_list:
+            logger.info("Load %s: %s",
+                        dashboard_utils.DashboardMasterModule.__name__, cls)
+            c = cls(self)
+            dashboard_utils.ClassMethodRouteTable.bind(c)
+            modules.append(c)
+        return modules
+
+    async def run(self):
+        # Create an aioredis client for all modules.
+        self.aioredis_client = await aioredis.create_redis_pool(
+            address=self.redis_address, password=self.redis_password)
+        # Waiting for GCS is ready.
+        while True:
+            try:
+                gcs_address = await self.aioredis_client.get(
+                    dashboard_consts.REDIS_KEY_GCS_SERVER_ADDRESS)
+                if not gcs_address:
+                    raise Exception("GCS address not found.")
+                channel = aiogrpc.insecure_channel(gcs_address)
+            except Exception as ex:
+                logger.error("Connect to GCS failed: %s, retry...", ex)
+                await asyncio.sleep(
+                    dashboard_consts.CONNECT_GCS_INTERVAL_SECONDS)
+            else:
+                self.aiogrpc_gcs_channel = channel
+                break
+
+        async def _async_notify():
```

Comment:
What is this for?

Reply:
This is for notifying async signals. Here is the usage in reporter_master.py: `DataSource.agents.signal.append(self._update_stubs)`

Reply:
So, it is used to notify each module that some data has been changed. Got it.